internal/statistics.go

```diff
@@ -0,0 +1,231 @@
+	"github.com/hazelcast/hazelcast-go-client/internal/util/versionutil"
+)
+
+const (
+	sinceVersionString = "3.9"
+	emptyStatValue     = ""
+	statSeperator      = ','
+	keyValueSeparator  = '='
+)
+
+var sinceVersion = versionutil.CalculateVersion(sinceVersionString)
+
+type statistics struct {
+	client       *HazelcastClient
+	enabled      bool
+	ownerAddress atomic.Value
+	gauges       map[string]func() string
+	cancel       chan struct{}
+}
+
+func newStatistics(client *HazelcastClient) *statistics {
+	stats := statistics{
+		client: client,
+		cancel: make(chan struct{}),
+	}
+	stats.enabled = client.properties.GetBoolean(property.StatisticsEnabled)
+	stats.gauges = make(map[string]func() string)
+	stats.ownerAddress.Store(&proto.Address{}) // initialize
+	return &stats
+}
+
+// start registers all client statistics and schedules periodic collection of stats.
+func (s *statistics) start() {
+	if !s.enabled {
+		return
+	}
+	s.registerMetrics()
+	period := s.client.properties.GetDuration(property.StatisticsPeriodSeconds)
+	if period <= 0 {
+		// This will return the default value since we have a non-positive value now
+		defaultPeriod := s.client.properties.GetPositiveDurationOrDef(property.StatisticsPeriodSeconds)
+		log.Println("Provided client statistics ", property.StatisticsPeriodSeconds.Name(),
+			" cannot be less than or equal to 0.",
+			" You provided ", period, " as the configuration. Client will use the default value ", defaultPeriod,
+			" instead.")
+		period = defaultPeriod
+	}
+	go s.schedulePeriodicStatisticsSendTask(period)
+}
+
+func (s *statistics) schedulePeriodicStatisticsSendTask(period time.Duration) {
+	s.sendStatistics()
+	ticker := time.NewTicker(period)
+	for {
+		select {
+		case <-s.cancel:
+			return
+		case <-ticker.C:
+			s.sendStatistics()
+		}
+	}
+}
+
+func (s *statistics) getOwnerConnection() *Connection {
+	connection := s.client.ConnectionManager.getOwnerConnection()
+	if connection == nil {
+		return nil
+	}
+	ownerConnectionAddress := s.client.ClusterService.getOwnerConnectionAddress()
+	serverVersion := connection.connectedServerVersion
+	if serverVersion < sinceVersion {
+		// do not print too many logs if connected to an old version server
+		cachedOwnerAddress, ok := s.ownerAddress.Load().(*proto.Address)
+		if !ok || (ownerConnectionAddress != cachedOwnerAddress) {
+			log.Println("Client statistics cannot be sent to server ", ownerConnectionAddress,
+				" since connected server version is less than the minimum supported server version ",
+				sinceVersionString)
+		}
+
+		// cache the last connected server address for decreasing the log prints
+		s.ownerAddress.Store(ownerConnectionAddress)
+		return nil
+	}
+	return connection
+}
+
+var emptyStatValueFunc = func() string {
+	return emptyStatValue
+}
+
+func (s *statistics) sendStatistics() {
+	ownerConnection := s.getOwnerConnection()
+	if ownerConnection == nil {
+		log.Println("Cannot send client statistics to the server. No owner connection")
```

Comment:
will this be printed periodically a number of times during which the client has no cluster connection since this method is called periodically? Is this desired?

Reply:
Currently this will be printed periodically. I am not sure if we should print this periodically, what do you think?

Reply:
Java impl:
            // suppress exception, do not print too many messages
            if (logger.isFinestEnabled()) {
                logger.finest("Could not send stats ", e);
            }

Reply:
Currently there is no logging level in the go client, it will be added in `0.5`. I think we can put the lowest level for this when it is implemented. What do you think?